### ISSUE TITLE
sql: TestLogic_cluster_locks fails with nil pointer dereference

### ISSUE THREAD

**Issue by rharding6373** [cockroachdb/cockroach#113935]
Found in unrelated PR https://github.com/cockroachdb/cockroach/pull/113527

TC: https://teamcity.cockroachdb.com/viewLog.html?buildId=12535137&buildTypeId=Cockroach_BazelEssentialCi

```
=== RUN   TestLogic_cluster_locks
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/0dbccfa3ae257c24fce6e73a3689e99e/logTestLogic_cluster_locks2318709849
    test_log_scope.go:81: use -show-logs to present logs inline
I231106 17:43:45.320848 1 (gostd) rand.go:250  [-] 1  random seed: 7705222062379041669
    logic.go:2966: let $root_session = 179519a40557a6810000000000000001
    logic.go:2966: let $testuser_session = 179519a41d7b5d1e0000000000000001
    logic.go:2966: let $txn1 = 0009ab4e-7fda-489c-998a-db0d93e58c2a
    logic.go:2966: let $txn2 = 589e476b-4945-4443-ae18-77089d4e6f45
    logic.go:2966: let $r1 = 67
    logic.go:2966: let $r2 = 68
    logic.go:2966: let $r3 = 69
    logic.go:2966: let $txn3 = 392f7976-139c-4294-90d4-3a1b78d90021
    logic.go:2966: let $txn4 = a049e81a-bc6a-4a5f-ada0-48c34dd745f0
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x358a415]

goroutine 3662 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0042d7730?, {0x6bb85d8, 0xc007416b40})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x65
panic({0x4d7f680?, 0x9c61710?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.populateClusterLocksWithFilter.genClusterLocksGenerator.func1.6()
	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:8022 +0x735
github.com/cockroachdb/cockroach/pkg/sql.populateClusterLocksWithFilter({0x6bb85d8?, 0xc007416b40?}, 0x6b74fa0?, {0x6c58218?, 0xc004c92bb0?}, 0xc00237d4d0, {0x0, 0x0, 0xc003bcc9e0, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:8058 +0xac
github.com/cockroachdb/cockroach/pkg/sql.init.genPopulateClusterLocksWithIndex.func117({0x6bb85d8, 0xc007416b40}, {0x6c18c20?, 0xc00726cc58}, 0x0?, {0x6c58218, 0xc004c92bb0}, 0x4?)
	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:8040 +0x11a
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).getPlanInfo.func1.(*virtualDefEntry).makeConstrainedRowsGenerator.func2({0x6bb85d8, 0xc007416b40}, {0x6b74fa0, 0xc00125e920})
	github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:823 +0x558
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3({0x6bb85d8, 0xc007416b40})
	github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:127 +0x102
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x13a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 2725
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x415
```

This is the line of code in crdb_internal.go that panics:
```
				tree.NewDString(tree.IsolationLevelFromKVTxnIsolationLevel(curLock.LockHolder.IsoLevel).String()), /* isolation_level */
```

I have not been able to repro this, even with the random seed.

Jira issue: CRDB-33263

**Comment by nvanbenschoten** [comment]
Something looks wrong in that code. There's an `if curLock.LockHolder != nil` check up above, but then we treat `curLock.LockHolder` as if it can't be nil in this (new) line of code:
```
tree.NewDString(tree.IsolationLevelFromKVTxnIsolationLevel(curLock.LockHolder.IsoLevel).String()), /* isolation_level */
```

**Comment by yuzefovich** [comment]
Yeah, it's a regression added in fa33c28abd95f5c0c586484b99dedc5597356a94, I'll fix it up.

**Comment by nvanbenschoten** [comment]
Thanks. The isolation level should be treated like `tsDatum`, `durationDatum`, and some of the other fields — it should come from either `curLock.LockHolder` or `curLock.Waiters[waiterIdx]`, depending on the position of the generator.

**Comment by yuzefovich** [comment]
Re-opening until #114670 is merged.

**Comment by yuzefovich** [comment]
Fixed by #114688
